validation: fix m_best_header tracking and BLOCK_FAILED_CHILD assignment

mzumsande opened the pull request:
`m_best_header` (the most-work header not known to be on an invalid chain) can be wrong in the context of invalidation / reconsideration of blocks. This can happen naturally (a valid header is received and stored in our block tree db; when the full block arrives, it is found to be invalid) or triggered by the user with the `invalidateblock` / `reconsiderblock` rpc.

We don't currently use `m_best_header` for any critical things (see OP of #16974 for a list that still seems up-to-date), so it being wrong affects mostly rpcs.

This PR proposes to recalculate it if necessary by looping over the block index and finding the best header. It also suggest to mark headers between an invalidatetd block and the previous `m_best_header` as invalid, so they won't be considered in the recalculation.
It adds tests to `rpc_invalidateblock.py` and `rpc_getchaintips.py` that fail on master.

One alternative to this suggested in the past would be to introduce a continuous tracking of header tips (#12138).
While this might be more performant, it is also more complicated, and situations where we need this data are only be remotely triggerable by paying the cost of creating a valid PoW header for an invalid block.
Therefore I think it isn't necessary to optimise for performance here, plus the solution in this PR doesn't perform any extra steps in the normal node operation where no invalidated blocks are encountered.

Fixes  #26245

mzumsande committed:
validation: mark blocks building on an invalid block as BLOCK_FAILED_CHILD

Without doing so, header-only chains building on a chain that
will be marked as invalid would still be eligible for m_best_header.
This improves both getblockchaininfo and getchaintips behavior.

While this adds an iteration over the entire block index, it can only be
triggered by the user (invalidateblock) or by others at a cost (the
header needs to be accepted in the first place, so it needs valid PoW).

Co-authored-by: TheCharlatan <seb.kung@gmail.com>

mzumsande committed:
validation: add RecalculateBestHeader() function

It recalculates m_best_header by looping over the entire
block index. Even though this is not very performant, it
will only be used in rare situations that cannot be
triggered by others without a cost:
As part of to invalidateblock / reconsiderblock rpcs, or when a
block with an accepted header with valid PoW turns out to be invalid
later during full validation.

mzumsande committed:
test: cleanup rpc_getchaintips.py

Remove whitespace that doesn't conform with pep8 and turn some
comments into log messages.

mzumsande committed:
rpc: call RecalculateBestHeader as part of reconsiderblock

Co-authored-by: Fabian Jahr <fjahr@protonmail.com>

mzumsande committed:
validation: call RecalculateBestHeader in InvalidChainFound

This means that it is being called in two situations:
1.) As part of the invalidateblock rpc
2.) When we receive a block for which we have a valid
header in our block index, but the block turns out to be invalid

mzumsande committed:
test: add test for getchaintips behavior with invalid chains

This test would fail to mark the chain tip as "invalid" instead
of "headers-only" without the previous commit marking the headers
as BLOCK_FAILED_CHILD.

fjahr reviewed (commented):
tACK c9bf06a531617fdd70b64e23572931af2b969828

I reviewed the code and confirmed manually that all the added tests do test the behavior changes in their commits.

I would also say this is a bugfix, so the title could be changed to "Fix m_best_header tracking" and get a bugfix label.

TheCharlatan commented on test/functional/rpc_invalidateblock.py:
```diff
@@ -35,9 +39,17 @@ def run_test(self):
+        self.nodes[0].submitheader(block.serialize().hex())
 
         self.log.info("Invalidate block 2 on node 0 and verify we reorg to node 0's original chain")
+        badhash = self.nodes[1].getblockhash(2)
```
It might be a good idea to expand this a bit:
```diff
@@ -45 +45 @@ class InvalidateTest(BitcoinTestFramework):
-        tip = int(self.nodes[0].getbestblockhash(), 16)
+        tip = self.nodes[0].getbestblockhash()
@@ -47 +47 @@ class InvalidateTest(BitcoinTestFramework):
-        block = create_block(tip, create_coinbase(self.nodes[0].getblockcount()), block_time, version=4)
+        block = create_block(int(tip, 16), create_coinbase(self.nodes[0].getblockcount()), block_time, version=4)
@@ -49,0 +50 @@ class InvalidateTest(BitcoinTestFramework):
+        assert_equal(self.nodes[0].getblockchaininfo()["headers"], self.nodes[0].getblockchaininfo()["blocks"] + 1)
@@ -56,0 +58,11 @@ class InvalidateTest(BitcoinTestFramework):
+        assert_equal(self.nodes[0].getblockchaininfo()["headers"], self.nodes[0].getblockchaininfo()["blocks"])
+
+        self.log.info("Reconsider block 6 on node 0 again and verify that the best header is set correctly")
+        self.nodes[0].reconsiderblock(tip)
+        assert_equal(self.nodes[0].getblockchaininfo()["headers"], self.nodes[0].getblockchaininfo()["blocks"] + 1)
+
+        self.log.info("Invalid block 2 on node 0 and verify we reorg to node 0's original chain again")
+        self.nodes[0].invalidateblock(badhash)
+        assert_equal(self.nodes[0].getblockcount(), 4)
+        assert_equal(self.nodes[0].getbestblockhash(), besthash_n0)
+        # Should report consistent blockchain info
```
This additionally checks that the new header can be reconsidered again correctly (`BLOCK_FAILED_CHILD` is indeed reset). It's a bit verbose and repetitive though and following along with the various chainstates is already a bit annoying, so not sure if it delivers enough value.

sedited reviewed (approved):
ACK c9bf06a531617fdd70b64e23572931af2b969828

TheCharlatan commented on src/validation.cpp:
```diff
@@ -6410,6 +6415,20 @@ std::optional<int> ChainstateManager::GetSnapshotBaseHeight() const
+        if (!(entry.second.nStatus & BLOCK_FAILED_MASK) && m_best_header->nChainWork < entry.second.nChainWork) {
+            m_best_header = &entry.second;
+        }
+    }
```
I've been looking the logic over again and think it might be a good idea to call `NotifyHeaderTip` here too. It basically tracks along `m_best_header`, but with the current code will only notify when processing the next block or header.

mzumsande replied:
Makes sense. I've added a call to `NotifyHeaderTip`.

mzumsande replied:
actually, that didn't work because `NotifyHeaderTip` can't be called with `cs_main` is held. Reverting this, I'll think about this a bit more.

TheCharlatan replied:
Ah right, I don't think this a blocker for this PR anyway, just something that came to my mind.

mzumsande replied:
I thought about it a bit and don't see a great non-invasive solution to this: We could add a call to `NotifyHeaderTip` in the calling functions after  `cs_main` is released, which would be in `ActivateBestChain()` (plus in the `invalidateblock` / `reconsiderblock` rpcs which seem less important), and I'm unsure if we want to refactor this function to always call `NotifyBestHeader()` just for the special case where a connected block turns out to be invalid.
In any case, since `m_best_header` is already being reset in `InvalidChainFound()` in master (just not always correctly), so this would not be a regression introduced by this PR, so I think it's ok not to address it here for now.

Yygik replied:
各方面看过嗯结果OK

mzumsande commented on src/validation.cpp:
```diff
@@ -2060,6 +2060,11 @@ void Chainstate::InvalidChainFound(CBlockIndex* pindexNew)
     if (m_chainman.m_best_header != nullptr && m_chainman.m_best_header->GetAncestor(pindexNew->nHeight) == pindexNew) {
+        CBlockIndex* index_walk{m_chainman.m_best_header};
+        while (index_walk != pindexNew) {
+            index_walk->nStatus |= BLOCK_FAILED_CHILD;
```
I think that this is not a complete fix. This would only mark indexes between `m_best_header` and the failed block as `BLOCK_FAILED_CHILD`, but if this was not  a linear chain and there were forked blocks, those wouldn't be marked as `BLOCK_FAILED_CHILD`. I think I'll change the logic to also set `BLOCK_FAILED_CHILD` while iterating over the entire block index.

mzumsande replied:
I've reworked this by introducing `SetBlockFailureFlags` (as a counterpart to `ResetBlockFailureFlags` that removes the flags after reconsiderblock)

mzumsande commented:
Thanks for the reviews so far! Ill address feedback soon, plus I want to rework 7d42b0e4c2722b398c73695d5547fbf8bd2ae01c, see comment above. I'm currently on holidays, so it might take a few days until I make the changes.

One more issue I've noticed while writing the tests this PR is that if we connect an invalid block received via p2p, `InvalidChainFound()` is called twice:
The first time in `ConnectTip()`, via `InvalidBlockFound()` [here](https://github.com/bitcoin/bitcoin/blob/c81c6bf65b34eb91c98768e6b4f46e450ccfd779/src/validation.cpp#L2072)  
The second time in `ActivateBestChainStep()` after `ConnectTip()` fails due to an invalid block ([here](https://github.com/bitcoin/bitcoin/blob/c81c6bf65b34eb91c98768e6b4f46e450ccfd779/src/validation.cpp#L3365)).
Does anyone know if there is a reason for this? It seems to me that the call in `ActivateBestChainStep()` is unnecessary and could be dropped.

luke-jr commented:
>It also suggest to mark headers between an invalidatetd block and the previous m_best_header as invalid, so they won't be considered in the recalculation.

You mean if you invalidateblock, then reconsiderblock, the rest of the child blocks also need to be reconsidered individually? That seems like a conceptual bug IMO, backward incompatible, and quite annoying.

mzumsande commented:
> You mean if you invalidateblock, then reconsiderblock, the rest of the child blocks also need to be reconsidered individually?

That's already happening, reconsiderblock already goes over the entire block index db on master, looking at each index individually, see `Chainstate::ResetBlockFailureFlags` (this also has a bug, #30479, but that is unrelated to the failure flags).

mzumsande commented on src/validation.cpp:
```diff
@@ -6410,6 +6410,20 @@ std::optional<int> ChainstateManager::GetSnapshotBaseHeight() const
+void ChainstateManager::RecalculateBestHeader()
+{
+    AssertLockHeld(cs_main);
+    // If, due to invalidation / reconsideration of blocks, the previous
```
done

mzumsande commented on test/functional/rpc_getchaintips.py:
```diff
@@ -58,5 +62,35 @@ def run_test(self):
+        n0.submitheader(block2.serialize().hex())
+        tips = n0.getchaintips()
+        assert_equal(len(tips), 3)
+        assert_equal(tips[0]['height'], start_height + 2)
```
Removed the line here and below

mzumsande commented:
[c9bf06a ](https://github.com/bitcoin/bitcoin/commit/c9bf06a531617fdd70b64e23572931af2b969828)to [5aa9160](https://github.com/bitcoin/bitcoin/commit/5aa9160698c4375afc8f242f8f1e3078e61b1713):
* rebased
* addressed feedback
* reworked commit f44a403ad27297c10256df11c3876b27c58dc1e4 (fix m_best_header tracking and BLOCK_FAILED_CHILD ) to set `BLOCK_FAILED_CHILD` for all blocks building on the invalid block, instead of just the ancestors of `m_best_header`

jonatack reviewed (commented):
Approach ACK 321097a685b67f82c32f53652227b7177fe97a95 on first look

fjahr commented:
> actually, that didn't work because NotifyHeaderTip can't be called when cs_main is held. Reverting this, I'll think about this a bit more.

Is this ready for review or should we wait until you have addressed this @mzumsande ?

mzumsande commented on src/validation.h:
```diff
@@ -1321,6 +1324,11 @@ class ChainstateManager
     std::optional<int> GetSnapshotBaseHeight() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
+    //! If, due to invalidation / reconsideration of blocks, the previous
+    //! best header is no longer valid or guaranteed to be the most-work
```
will change to "/" so it mirrors the "invalidation / reconsideration" in the line before.

mzumsande commented on src/validation.h:
```diff
@@ -735,6 +735,9 @@ class Chainstate
         LOCKS_EXCLUDED(::cs_main);
 
+    /** Set invalidity status to a all descendants of a block */
+    void SetBlockFailureFlags(CBlockIndex* pindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
change to `::cs_main`

mzumsande commented:
[321097a ](https://github.com/bitcoin/bitcoin/commit/321097a685b67f82c32f53652227b7177fe97a95)to [0bd53d9](https://github.com/bitcoin/bitcoin/commit/0bd53d913c1c2ffd2d0779f01bc51c81537b6992): addressed feedback by @jonatack

>Is this ready for review or should we wait until you have addressed this @mzumsande ?

It's ready for review, see https://github.com/bitcoin/bitcoin/pull/30666#discussion_r1763467219 - but of course I'm open to suggestions!

fjahr commented:
reACK 0bd53d913c1c2ffd2d0779f01bc51c81537b6992

sedited reviewed (approved):
Re-ACK 0bd53d913c1c2ffd2d0779f01bc51c81537b6992

achow101 commented:
ACK 0bd53d913c1c2ffd2d0779f01bc51c81537b6992

Yygik commented:
i额迷你发光你放哪更好
